triqs/master segfaults clang-4.0.1

Hi,

I just upgraded to clang-4.0.1 and `triqs/master` breaks clang with a segfault. I am trying to reproduce this for `unstable` as well. Please see below for the error message.

Best, Hugo

```
[ 42%] Building CXX object test/triqs/gfs/CMakeFiles/g_r_k.dir/g_r_k.cpp.o
In file included from /Users/hugstr/dev/triqs/test/triqs/gfs/gf_mul.cpp:1:
In file included from /Users/hugstr/dev/triqs/triqs/test_tools/gfs.hpp:3:
In file included from /Users/hugstr/dev/triqs/triqs/./gfs/../lattice/../gfs.hpp:39:
/Users/hugstr/dev/triqs/triqs/gfs/block.hpp:147:2: error: too many template arguments
      for class template 'gf'
 block_gf<A...> make_block_gf(std::vector<std::string> block_names, std::initial...
 ^~~~~~~~
/Users/hugstr/dev/triqs/test/triqs/gfs/gf_mul.cpp:10:11: note: in instantiation of
      function template specialization 'triqs::gfs::make_block_gf<triqs::gfs::imfreq,
      triqs::gfs::matrix_valued, triqs::gfs::tail, void>' requested here
 auto A = make_block_gf<imfreq>({"up", "down"}, {gf<imfreq>{{beta, Fermion}, {1...
          ^
/Users/hugstr/dev/triqs/triqs/./gfs/./impl/gf.hpp:28:8: note: template is declared here
 class gf : TRIQS_CONCEPT_TAG_NAME(ImmutableGreenFunction) {
       ^
0  libLLVM.dylib            0x000000010e035cb5 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 37
1  libLLVM.dylib            0x000000010e03534f llvm::sys::RunSignalHandlers() + 83
2  libLLVM.dylib            0x000000010e036092 SignalHandler(int) + 182
3  libsystem_platform.dylib 0x00007fffc532ab3a _sigtramp + 26
4  libsystem_platform.dylib 0x00000001120334a0 _sigtramp + 1288735104
5  clang                    0x000000010b1538a2 clang::Sema::FinishTemplateArgumentDeduction(clang::FunctionTemplateDecl*, llvm::SmallVectorImpl<clang::DeducedTemplateArgument>&, unsigned int, clang::FunctionDecl*&, clang::sema::TemplateDeductionInfo&, llvm::SmallVectorImpl<clang::Sema::OriginalCallArg> const*, bool, llvm::function_ref<bool ()>) + 2818
6  clang                    0x000000010b15447a clang::Sema::DeduceTemplateArguments(clang::FunctionTemplateDecl*, clang::TemplateArgumentListInfo*, llvm::ArrayRef<clang::Expr*>, clang::FunctionDecl*&, clang::sema::TemplateDeductionInfo&, bool, llvm::function_ref<bool (llvm::ArrayRef<clang::QualType>)>) + 1544
7  clang                    0x000000010b0dda51 clang::Sema::AddTemplateOverloadCandidate(clang::FunctionTemplateDecl*, clang::DeclAccessPair, clang::TemplateArgumentListInfo*, llvm::ArrayRef<clang::Expr*>, clang::OverloadCandidateSet&, bool, bool) + 405
8  clang                    0x000000010b0e7967 AddOverloadedCallCandidate(clang::Sema&, clang::DeclAccessPair, clang::TemplateArgumentListInfo*, llvm::ArrayRef<clang::Expr*>, clang::OverloadCandidateSet&, bool, bool) + 175
9  clang                    0x000000010b0e77f2 clang::Sema::AddOverloadedCallCandidates(clang::UnresolvedLookupExpr*, llvm::ArrayRef<clang::Expr*>, clang::OverloadCandidateSet&, bool) + 252
10 clang                    0x000000010b0e7a0f clang::Sema::buildOverloadedCallSet(clang::Scope*, clang::Expr*, clang::UnresolvedLookupExpr*, llvm::MutableArrayRef<clang::Expr*>, clang::SourceLocation, clang::OverloadCandidateSet*, clang::ActionResult<clang::Expr*, true>*) + 161
11 clang                    0x000000010b0e7c65 clang::Sema::BuildOverloadedCallExpr(clang::Scope*, clang::Expr*, clang::UnresolvedLookupExpr*, clang::SourceLocation, llvm::MutableArrayRef<clang::Expr*>, clang::SourceLocation, clang::Expr*, bool, bool) + 237
12 clang                    0x000000010afc734b clang::Sema::ActOnCallExpr(clang::Scope*, clang::Expr*, clang::SourceLocation, llvm::MutableArrayRef<clang::Expr*>, clang::SourceLocation, clang::Expr*, bool) + 451
13 clang                    0x000000010ad9735d clang::Parser::ParsePostfixExpressionSuffix(clang::ActionResult<clang::Expr*, true>) + 2997
14 clang                    0x000000010ad98c37 clang::Parser::ParseCastExpression(bool, bool, bool&, clang::Parser::TypeCastState) + 4297
15 clang                    0x000000010ad96710 clang::Parser::ParseCastExpression(bool, bool, clang::Parser::TypeCastState) + 40
16 clang                    0x000000010ad9a07d clang::Parser::ParseCastExpression(bool, bool, bool&, clang::Parser::TypeCastState) + 9487
17 clang                    0x000000010ad96710 clang::Parser::ParseCastExpression(bool, bool, clang::Parser::TypeCastState) + 40
18 clang                    0x000000010ad95a89 clang::Parser::ParseAssignmentExpression(clang::Parser::TypeCastState) + 149
19 clang                    0x000000010ad7d0ee clang::Parser::ParseDeclarationAfterDeclaratorAndAttributes(clang::Declarator&, clang::Parser::ParsedTemplateInfo const&, clang::Parser::ForRangeInit*) + 2454
20 clang                    0x000000010ad7b95c clang::Parser::ParseDeclGroup(clang::ParsingDeclSpec&, unsigned int, clang::SourceLocation*, clang::Parser::ForRangeInit*) + 1724
21 clang                    0x000000010ad7875f clang::Parser::ParseSimpleDeclaration(unsigned int, clang::SourceLocation&, clang::Parser::ParsedAttributesWithRange&, bool, clang::Parser::ForRangeInit*) + 613
22 clang                    0x000000010ad7849f clang::Parser::ParseDeclaration(unsigned int, clang::SourceLocation&, clang::Parser::ParsedAttributesWithRange&) + 373
23 clang                    0x000000010adc4009 clang::Parser::ParseStatementOrDeclarationAfterAttributes(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::AllowedContsructsKind, clang::SourceLocation*, clang::Parser::ParsedAttributesWithRange&) + 507
24 clang                    0x000000010adc3d8e clang::Parser::ParseStatementOrDeclaration(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::AllowedContsructsKind, clang::SourceLocation*) + 142
25 clang                    0x000000010adc89c0 clang::Parser::ParseCompoundStatementBody(bool) + 1228
26 clang                    0x000000010adc90bf clang::Parser::ParseFunctionStatementBody(clang::Decl*, clang::Parser::ParseScope&) + 165
27 clang                    0x000000010add7af1 clang::Parser::ParseFunctionDefinition(clang::ParsingDeclarator&, clang::Parser::ParsedTemplateInfo const&, clang::Parser::LateParsedAttrList*) + 2605
28 clang                    0x000000010ad7b870 clang::Parser::ParseDeclGroup(clang::ParsingDeclSpec&, unsigned int, clang::SourceLocation*, clang::Parser::ForRangeInit*) + 1488
29 clang                    0x000000010add6fe4 clang::Parser::ParseDeclOrFunctionDefInternal(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec&, clang::AccessSpecifier) + 800
30 clang                    0x000000010add6b3d clang::Parser::ParseDeclarationOrFunctionDefinition(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*, clang::AccessSpecifier) + 251
31 clang                    0x000000010add6396 clang::Parser::ParseExternalDeclaration(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*) + 2088
32 clang                    0x000000010add59db clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&) + 433
33 clang                    0x000000010ad6e580 clang::ParseAST(clang::Sema&, bool, bool) + 326
34 clang                    0x000000010abe0ccb clang::FrontendAction::Execute() + 67
35 clang                    0x000000010abb47f4 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) + 714
36 clang                    0x000000010ac0adde clang::ExecuteCompilerInvocation(clang::CompilerInstance*) + 3786
37 clang                    0x000000010a9158c7 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) + 1203
38 clang                    0x000000010a914b6c main + 7636
39 libdyld.dylib            0x00007fffc511b235 start + 1
Stack dump:
0.	Program arguments: /opt/local/libexec/llvm-4.0/bin/clang -cc1 -triple x86_64-apple-macosx10.12.0 -Wdeprecated-objc-isa-usage -Werror=deprecated-objc-isa-usage -emit-obj -disable-free -disable-llvm-verifier -discard-value-names -main-file-name gf_mul.cpp -mrelocation-model pic -pic-level 2 -mthread-model posix -mdisable-fp-elim -menable-no-infs -menable-no-nans -menable-unsafe-fp-math -fno-signed-zeros -freciprocal-math -fno-trapping-math -ffp-contract=fast -ffast-math -masm-verbose -munwind-tables -target-cpu skylake -target-feature +sse2 -target-feature +cx16 -target-feature -tbm -target-feature -avx512ifma -target-feature -avx512dq -target-feature -fma4 -target-feature +prfchw -target-feature +bmi2 -target-feature +xsavec -target-feature +fsgsbase -target-feature +popcnt -target-feature +aes -target-feature -pcommit -target-feature +xsaves -target-feature -avx512er -target-feature -clwb -target-feature -avx512f -target-feature -pku -target-feature +smap -target-feature +mmx -target-feature -xop -target-feature +rdseed -target-feature -hle -target-feature -sse4a -target-feature -avx512bw -target-feature +clflushopt -target-feature +xsave -target-feature -avx512vl -target-feature +invpcid -target-feature -avx512cd -target-feature +avx -target-feature -rtm -target-feature +fma -target-feature +bmi -target-feature -mwaitx -target-feature +rdrnd -target-feature +sse4.1 -target-feature +sse4.2 -target-feature +avx2 -target-feature +sse -target-feature +lzcnt -target-feature +pclmul -target-feature -prefetchwt1 -target-feature +f16c -target-feature +ssse3 -target-feature +sgx -target-feature +cmov -target-feature -avx512vbmi -target-feature +movbe -target-feature +xsaveopt -target-feature -sha -target-feature +adx -target-feature -avx512pf -target-feature +sse3 -target-linker-version 274.2 -dwarf-column-info -debug-info-kind=standalone -dwarf-version=4 -debugger-tuning=gdb -coverage-notes-file /Users/hugstr/dev/triqs/cbuild/test/triqs/gfs/CMakeFiles/gf_mul.dir/gf_mul.cpp.gcno -resource-dir /opt/local/libexec/llvm-4.0/bin/../lib/clang/4.0.1 -isystem /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/numpy/core/include -isystem /opt/local/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7 -isystem /opt/local/include/openmpi-clang40 -isystem /opt/local/include -D HAVE_NFFT -D TRIQS_BUILDING_LIBRARY -I /Users/hugstr/dev/triqs -I /Users/hugstr/dev/triqs/cbuild/include -I /Users/hugstr/dev/triqs/cbuild/Config -I /opt/local/include/openmpi-clang40 -stdlib=libc++ -stdlib=libc++ -Ofast -Wno-deprecated-writable-strings -std=c++1y -fdeprecated-macro -fdebug-compilation-dir /Users/hugstr/dev/triqs/cbuild/test/triqs/gfs -ferror-limit 19 -fmessage-length 88 -I/opt/local/include/libomp -fopenmp -pthread -stack-protector 1 -fblocks -fobjc-runtime=macosx-10.12.0 -fencode-extended-block-signature -fcxx-exceptions -fexceptions -fmax-type-align=16 -fdiagnostics-show-option -fcolor-diagnostics -vectorize-loops -vectorize-slp -o CMakeFiles/gf_mul.dir/gf_mul.cpp.o -x c++ /Users/hugstr/dev/triqs/test/triqs/gfs/gf_mul.cpp 
1.	/Users/hugstr/dev/triqs/test/triqs/gfs/gf_mul.cpp:10:123: current parser token ')'
2.	/Users/hugstr/dev/triqs/test/triqs/gfs/gf_mul.cpp:6:27: parsing function body 'BlockGf_Operations_Test::TestBody'
3.	/Users/hugstr/dev/triqs/test/triqs/gfs/gf_mul.cpp:6:27: in compound statement ('{}')
clang: error: unable to execute command: Segmentation fault: 11
clang: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 4.0.1 (tags/RELEASE_401/final)
Target: x86_64-apple-darwin16.7.0
Thread model: posix
InstalledDir: /opt/local/libexec/llvm-4.0/bin
clang: note: diagnostic msg: PLEASE submit a bug report to http://llvm.org/bugs/ and include the crash backtrace, preprocessed source, and associated run script.
clang: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /var/folders/p9/bgyyj82s2vzd0k7swf3c6wwm0000gn/T/gf_mul-bfe342.cpp
clang: note: diagnostic msg: /var/folders/p9/bgyyj82s2vzd0k7swf3c6wwm0000gn/T/gf_mul-bfe342.sh
clang: note: diagnostic msg: Crash backtrace is located in
clang: note: diagnostic msg: /Users/hugstr/Library/Logs/DiagnosticReports/clang_<YYYY-MM-DD-HHMMSS>_<hostname>.crash
clang: note: diagnostic msg: (choose the .crash file that corresponds to your crash)
clang: note: diagnostic msg: 

********************
```

--- a/test/triqs/gfs/block.cpp
+++ b/test/triqs/gfs/block.cpp
@@ -17,9 +17,9 @@ TEST(Gf, Block) {
   auto B01 = block_gf<imfreq>(v);
   auto B00 = block_gf<imfreq>({"a","b","c"});
 
-  auto B1 = make_block_gf<imfreq>(3, G1);
-  auto B2 = make_block_gf<imfreq>({G1, G1, G1});
-  auto B3 = make_block_gf<imfreq>({"a", "b", "c"}, {G1, G1, G1});
+  auto B1 = make_block_gf(3, G1);
+  auto B2 = make_block_gf({G1, G1, G1});
+  auto B3 = make_block_gf({"a", "b", "c"}, {G1, G1, G1});
   auto B4 = block_gf<imfreq>(1);
   //TEST(B3["a"]); //does not compile
 
@@ -79,7 +79,7 @@ TEST(Gf, Block) {
   // inversion
   {
    auto inv_G1 = inverse(G1);
-   auto B = make_block_gf<imfreq>(3, G1);
+   auto B = make_block_gf(3, G1);
    auto inv_B1 = inverse(B);
    for (auto & g : inv_B1) EXPECT_GF_NEAR(g,inv_G1);
   }
@@ -94,8 +94,8 @@ TEST(Gf, Block) {
   G1(w_) << 1 / (w_ + 2);
   G2(w_) << 1 / (w_ - 2);
 
-  auto B_ab = make_block_gf<imfreq>({"a", "b"}, {G1, G2});
-  auto B_ba = make_block_gf<imfreq>({"b", "a"}, {G2, G1});
+  auto B_ab = make_block_gf({"a", "b"}, {G1, G2});
+  auto B_ba = make_block_gf({"b", "a"}, {G2, G1});
   
   auto B_res = B_ab;
   for(auto const & ind : B_res.mesh()){

--- a/test/triqs/gfs/gf_mul.cpp
+++ b/test/triqs/gfs/gf_mul.cpp
@@ -7,7 +7,7 @@ TEST(BlockGf, Operations) {
 
  double beta = 10.0;
 
- auto A = make_block_gf<imfreq>({"up", "down"}, {gf<imfreq>{{beta, Fermion}, {1, 1}}, gf<imfreq>{{beta, Fermion}, {1, 1}}});
+ auto A = make_block_gf({"up", "down"}, {gf<imfreq>{{beta, Fermion}, {1, 1}}, gf<imfreq>{{beta, Fermion}, {1, 1}}});
  auto B = A;
  auto C = A;
 

--- a/test/triqs/gfs/multivar/g_nu_nup.cpp
+++ b/test/triqs/gfs/multivar/g_nu_nup.cpp
@@ -45,7 +45,7 @@ TEST(BlockGfCartesian, H5_RW_Evaluator){
  double beta=1;
  auto g = gf<cartesian_product<imfreq, imfreq>, matrix_valued>{{{beta, Fermion, 5}, {beta, Boson, 5}}, {1,1}};
  g()=2;
- auto G = make_block_gf<cartesian_product<imfreq, imfreq>, matrix_valued>({"up"}, {g});
+ auto G = make_block_gf({"up"}, {g});
  //EXPECT_ARRAY_NEAR(get_target_shape(G[0]), mini_vector<size_t,2>{1,1});
  h5::file file("g_nu_nup.h5", H5F_ACC_TRUNC );
  h5_write(file, "G", G);
@@ -74,7 +74,7 @@ TEST(BlockGfCartesian, OutOfBounds){
  }
 
  //g_2w.singularity().toto(); //tail_zero
- auto G_2w = make_block_gf<cartesian_product<imfreq, imfreq>, tensor_valued<3>>({"0"}, {g_2w});
+ auto G_2w = make_block_gf({"0"}, {g_2w});
  //G_2w[0].singularity().toto();//tail_zero
  //G_2w.singularity().toto();//nothing
  std::cout << evaluate(G_2w[0].singularity(), 5) << std::endl;
@@ -106,7 +106,7 @@ TEST(BlockGf, H5_RW_Evaluator){
 
  auto g = gf<imfreq, matrix_valued>{{beta, Boson, 5}, {1,1}};
  g()=2;
- auto G = make_block_gf<imfreq, matrix_valued>({"up"}, {g});
+ auto G = make_block_gf({"up"}, {g});
 
 
  h5::file file("g_nu_nup.h5", H5F_ACC_TRUNC );

--- a/test/triqs/gfs/multivar/mpi.cpp
+++ b/test/triqs/gfs/multivar/mpi.cpp
@@ -34,9 +34,9 @@ TEST(Gfs, MPI_multivar) {
  // Test the result ?
 
 
- auto G = make_block_gf<cartesian_product<imfreq, imfreq, imfreq>, scalar_valued>({g});
+ auto G = make_block_gf({g});
  auto g0 = gf<imfreq, scalar_valued>{{beta, Boson, nbw}};
- auto G2 = make_block_gf<imfreq, scalar_valued>({g0});
+ auto G2 = make_block_gf({g0});
 
  mpi_broadcast(G, world);
  mpi_broadcast(G2, world);

